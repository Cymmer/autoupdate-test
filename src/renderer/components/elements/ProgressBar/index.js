import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import styles from './styles.module.scss';

const ProgressBar = ({ target }) => {
  const [readingProgress, setReadingProgress] = useState(0);

  // Source : https://nehalist.io/creating-a-reading-progress-bar-in-react/

  const scrollListener = () => {
    if (!target.current) {
      return;
    }

    const element = target.current;
    const totalHeight =
      element.clientHeight - element.offsetTop - window.innerHeight;
    let windowScrollTop =
      window.pageYOffset ||
      document.documentElement.scrollTop ||
      document.body.scrollTop ||
      0;

    // we subtract 276 as the offset value from the top
    windowScrollTop -= 276;

    if (windowScrollTop === 0) {
      return setReadingProgress(0);
    }

    if (windowScrollTop > totalHeight) {
      return setReadingProgress(100);
    }

    setReadingProgress((windowScrollTop / totalHeight) * 100);
  };

  useEffect(() => {
    window.addEventListener('scroll', scrollListener);
    return () => window.removeEventListener('scroll', scrollListener);
  });

  return (
    <progress
      className={styles.ProgressBar}
      max={100}
      value={readingProgress}
    />
  );
};

ProgressBar.propTypes = {
  target: PropTypes.object.isRequired,
};

export default ProgressBar;
