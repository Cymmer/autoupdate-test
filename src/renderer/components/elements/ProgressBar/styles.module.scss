@import '../../../scss/variables';
@import '../../../scss/mixins';
@import '../../../scss/themes';

.ProgressBar {
  /* Reset the default appearance */
  appearance: none;

  /* Get rid of default border in Firefox. */
  border: none;

  &::-webkit-progress-bar {
    @include theme {
      background-color: t('textColor-100');
    }
  }

  &::-webkit-progress-value {
    background: linear-gradient(90deg, #2fa9e0 0%, #65d0b7 100%);
    transition: width 0.4s ease;
    border-radius: 0 8px 8px 0;
  }

  /* For Mozilla */
  &::-moz-progress-bar {
    background: linear-gradient(90deg, #2fa9e0 0%, #65d0b7 100%);
    transition: width 0.4s ease;
    border-radius: 0 8px 8px 0;
  }

  /* For IE10 */
  color: linear-gradient(90deg, #2fa9e0 0%, #65d0b7 100%);

  width: 100%;
  height: 10px;
  justify-self: center;
  position: fixed;
  z-index: 4;
}
