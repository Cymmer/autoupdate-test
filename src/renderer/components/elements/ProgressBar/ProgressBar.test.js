import React from 'react';
import { shallow } from 'enzyme';
import ProgressBar from './index';

describe('RENDERING - ProgressBar', () => {
  it('should render with default props', () => {
    const component = shallow(
      <ProgressBar
        target={{
          current: {},
        }}
      />
    );
    expect(component).toMatchSnapshot();
  });

  // TODO: Implement appropriate testing for scrolling behavior
});
