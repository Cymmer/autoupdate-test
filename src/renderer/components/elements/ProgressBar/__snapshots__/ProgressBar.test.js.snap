// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RENDERING - ProgressBar should render with default props 1`] = `
<progress
  className="ProgressBar"
  max={100}
  value={0}
/>
`;
